kobuki_node/src/kobuki_ros.cpp

```diff
@@ -0,0 +1,972 @@
+    kobuki_.disable();
+    odometry_->resetTimeout(this->get_clock()->now());
+  }
+  else {
+    RCLCPP_ERROR(get_logger(), "Kobuki : Motor power command specifies unknown state '%u'.", (unsigned int)msg->state);
+  }
+}
+
+void KobukiRos::subscribeControllerInfoCommand(const std::shared_ptr<kobuki_ros_interfaces::msg::ControllerInfo> msg)
+{
+  if (msg->p_gain < 0.0f ||  msg->i_gain < 0.0f ||  msg->d_gain < 0.0f) {
+    RCLCPP_ERROR(get_logger(), "Kobuki : All controller gains should be positive.");
+    return;
+  }
+  kobuki_.setControllerGain(msg->type,
+                            static_cast<unsigned int>(msg->p_gain*1000.0f),
+                            static_cast<unsigned int>(msg->i_gain*1000.0f),
+                            static_cast<unsigned int>(msg->d_gain*1000.0f));
+}
+
+void KobukiRos::processStreamData()
+{
+  publishWheelState();
+  publishSensorState();
+  publishDockIRData();
+  publishInertia();
+  publishRawInertia();
+}
+
+/*****************************************************************************
+** Publish Sensor Stream Workers
+*****************************************************************************/
+
+void KobukiRos::publishSensorState()
+{
+  kobuki_ros_interfaces::msg::SensorState state;
+  kobuki::CoreSensors::Data data = kobuki_.getCoreSensorData();
+  state.header.stamp = this->get_clock()->now();
+  state.time_stamp = data.time_stamp; // firmware time stamp
+  state.bumper = data.bumper;
+  state.wheel_drop = data.wheel_drop;
+  state.cliff = data.cliff;
+  state.left_encoder = data.left_encoder;
+  state.right_encoder = data.right_encoder;
+  state.left_pwm = data.left_pwm;
+  state.right_pwm = data.right_pwm;
+  state.buttons = data.buttons;
+  state.charger = data.charger;
+  state.battery = data.battery;
+  state.over_current = data.over_current;
+
+  kobuki::Cliff::Data cliff_data = kobuki_.getCliffData();
+  state.bottom = cliff_data.bottom;
+
+  kobuki::Current::Data current_data = kobuki_.getCurrentData();
+  state.current = current_data.current;
+
+  kobuki::GpInput::Data gp_input_data = kobuki_.getGpInputData();
+  state.digital_input = gp_input_data.digital_input;
+  for (unsigned int i = 0; i < gp_input_data.analog_input.size(); ++i) {
+    state.analog_input.push_back(gp_input_data.analog_input[i]);
+  }
+
+  sensor_state_publisher_->publish(state);
+}
+
+void KobukiRos::publishWheelState()
+{
+  // Take latest encoders and gyro data
+  ecl::LegacyPose2D<double> pose_update;
+  ecl::linear_algebra::Vector3d pose_update_rates;
+  kobuki_.updateOdometry(pose_update, pose_update_rates);
+  kobuki_.getWheelJointStates(joint_states_.position[0], joint_states_.velocity[0],   // left wheel
+                              joint_states_.position[1], joint_states_.velocity[1]);  // right wheel
+
+  // Update and publish odometry and joint states
+  geometry_msgs::msg::Quaternion odom_quat = odometry_->update(pose_update, pose_update_rates, kobuki_.getHeading(), kobuki_.getAngularVelocity());
```

Comment:
This is a bit awkward - lots of tro and fro with variables, partial state getting stored and simple getters becoming complex. I think what you were trying to accomplish here was to get all of the publishers and broadcasters back in the `KobukiNode` class. To do that, and make this less complicated, perhaps:

* Don't return anything on `odometry_->update`, store relevant (and complete, not partial) state in `odometry_`. 
* Pass in the timestamp on `odometry_->update()` and store it as well. i.e. again, make sure information in `odometry_` is *complete*.
* Then `getTransform()` and `getOdometry()` take no arguments and are simply about converting and retrieving the state as msg information

Reply:
Good idea.  I switched to this style in 9605787, and I think it is nicer.  Take a look and let me know what you think.